pkg/core/native/native_neo.go

```diff
@@ -317,15 +282,15 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pubs keys.Public
 
 // ModifyAccountVotes modifies votes of the specified account by value (can be negative).
 func (n *NEO) ModifyAccountVotes(acc *state.NEOBalanceState, d dao.DAO, value *big.Int) error {
-	for _, vote := range acc.Votes {
-		key := makeValidatorKey(vote)
+	if acc.VoteTo != nil {
+		key := makeValidatorKey(acc.VoteTo)
 		si := d.GetStorageItem(n.ContractID, key)
 		if si == nil {
 			return errors.New("invalid validator")
 		}
-		votes := bigint.FromBytes(si.Value)
-		votes.Add(votes, value)
-		si.Value = bigint.ToPreallocatedBytes(votes, si.Value[:0])
+		cd := new(candidate).FromBytes(si.Value)
```

Comment:
We need to check `cd.Registered` somewhere here.

Reply:
Still relevant, it shouldn't be possible to vote for unregistered candidates. Although we should also be careful to not break balance changing for accounts that voted for previously registered candidates, so this check actually has to be done in `VoteInternal`.

Reply:
No, because this is called in `Transfer` and they shouldn't be blocked because someone is unregistered.
But we still have another issue:
1. A votes for C
2. A has zero balance
3. C is unregistered (and thus deleted)
4. A does some transfer. <- **will fail**
This is how it is done in C#.

Reply:
```
        private bool Vote(ApplicationEngine engine, UInt160 account, ECPoint voteTo)
...
           if (voteTo != null)
            {
                validator_new = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Candidate).Add(voteTo))?.GetInteroperable<CandidateState>();
                if (validator_new is null) return false;
                if (!validator_new.Registered) return false;
            }
```
> it shouldn't be possible to vote for unregistered candidates

> this check actually has to be done in VoteInternal

So, we need to do this check when voting.

> But we still have another issue:

Is it really possible? I don't think A can have a valid balance record with zero value.

Reply:
@roman-khimov I see now. Haven't updated for a while on my local machine.